util/cmp/stream.go

```diff
@@ -0,0 +1,234 @@
+		return fmt.Errorf("error sending app files to cmp-server: %s", err)
+	}
+	return nil
+}
+
+// sendFile will send the file over the gRPC stream using a
+// buffer.
+func sendFile(ctx context.Context, sender StreamSender, file *os.File) error {
+	reader := bufio.NewReader(file)
+	chunk := make([]byte, 1024)
+	for {
+		if ctx != nil {
+			if err := ctx.Err(); err != nil {
+				return fmt.Errorf("client stream context error: %s", err)
+			}
+		}
+		n, err := reader.Read(chunk)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("buffer reader error: %s", err)
+		}
+		fr := appFileRequest(chunk[:n])
+		err = sender.Send(fr)
+		if err != nil {
+			return fmt.Errorf("error sending stream: %s", err)
+		}
+	}
+	return nil
+}
+
+// compressFiles will create a tgz file with all contents of appPath
+// directory excluding the .git folder. Returns the file alongside
+// its sha256 hash to be used as checksum. It is the resposibility
+// of the caller to close the file.
+func compressFiles(appPath string) (*os.File, string, error) {
+	excluded := []string{".git"}
+	appName := filepath.Base(appPath)
+	tgzFile, err := ioutil.TempFile(os.TempDir(), appName)
+	if err != nil {
+		return nil, "", fmt.Errorf("error creating app temp tgz file: %s", err)
+	}
+	hasher := sha256.New()
+	err = files.Tgz(appPath, excluded, tgzFile, hasher)
+	if err != nil {
+		tgzFile.Close()
+		os.Remove(tgzFile.Name())
+		return nil, "", fmt.Errorf("error creating app tgz file: %s", err)
+	}
+	checksum := hex.EncodeToString(hasher.Sum(nil))
+
+	// reposition the offset to the beginning of the file for proper reads
+	_, err = tgzFile.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, "", fmt.Errorf("error processing tgz file: %s", err)
+	}
+	return tgzFile, checksum, nil
+}
+
+// receiveFile will receive the file from the gRPC stream and save it in the dst folder.
+// Returns error if checksum doesn't match the one provided in the fileMetadata.
+// It is responsibility of the caller to close the returned file.
+func receiveFile(ctx context.Context, receiver StreamReceiver, checksum, dst string) (*os.File, error) {
+	fileBuffer := bytes.Buffer{}
+	hasher := sha256.New()
+	for {
+		if ctx != nil {
+			if err := ctx.Err(); err != nil {
+				return nil, fmt.Errorf("stream context error: %s", err)
+			}
+		}
+		req, err := receiver.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, fmt.Errorf("stream Recv error: %s", err)
+		}
+		f := req.GetFile()
+		if f == nil {
+			return nil, fmt.Errorf("stream request file is nil")
+		}
+		_, err = fileBuffer.Write(f.Chunk)
+		if err != nil {
+			return nil, fmt.Errorf("error writing file buffer: %s", err)
+		}
+		_, err = hasher.Write(f.Chunk)
+		if err != nil {
+			return nil, fmt.Errorf("error writing hasher: %s", err)
+		}
+	}
+	if hex.EncodeToString(hasher.Sum(nil)) != checksum {
+		return nil, fmt.Errorf("file checksum validation error")
+	}
+
+	file, err := ioutil.TempFile(dst, "")
```

Comment:
We should probably use a path containing a UUIDv4.

Reply:
`dst` is a function parameter that is already pointing to a UUIDv4 named directory.